File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int>st;
        int n1 = nums1.size();
        int n2 = nums2.size();
        unordered_map<int,int>mpp;
        for(int num : nums2){
            while(!st.empty() && num > st.top()){
                mpp[st.top()] = num;
                st.pop(); // remove the last element from the stack
            }
            st.push(num);
        }
        while(!st.empty()){
            mpp[st.top()] = -1;
            st.pop();
        }
        vector<int>res;
        for(int num : nums1){
            res.push_back(mpp[num]);
        }
        return res;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n1 = nums1.length;
        int n2 = nums2.length;
        Stack<Integer>st = new Stack<>();
        Map<Integer, Integer>mpp = new HashMap<>();
        
        // Store them into stack, According to the next greater element..
        for(int num : nums2){
            while(!st.isEmpty() && num > st.peek()){
                mpp.put(st.pop(), num);
            }
            st.push(num);
        }
        // Rest of the elements, assign them to "-1"
        while(!st.isEmpty()){
            mpp.put(st.pop(), -1);
        }

        int[] result = new int[n1];
        Arrays.fill(result, -1);
        for(int i = 0; i < n1; i++){
            result[i] = mpp.get(nums1[i]);  // "Get" automatically calls the value{key, value}
        }
        
        return result;
    }
}